contracts/Nodes.sol

```diff
@@ -139,32 +146,47 @@ contract Nodes is AccessManagedUpgradeable, INodes {
 
     function registerNode(
         bytes calldata ip,
+        bytes32[2] calldata publicKey,
         uint16 port
     )
         external
         override
         validIp(ip)
         validPort(port)
+        validPubKey(publicKey)
     {
-        NodeId nodeId = _createActiveNode(msg.sender, ip, port, "");
+        require(
+            msg.sender == _publicKeyToAddress(publicKey),
+            InvalidPublicKeyForSender(publicKey, _publicKeyToAddress(publicKey), msg.sender)
```

Comment:
In this case `_publicKeyToAddress` is calculated twice even when msg.sender is correct.